features/crc/soft/FastCRC.h

```diff
@@ -0,0 +1,202 @@
+                _width = 7;
+                break;
+
+            default:
+            case CRC_8BIT_CCITT:
+                _polynomial = 0x07;            // x8+x2+x+1
+                _width = 8;
+                break;
+
+            case CRC_16BIT_SD:
+                _polynomial = 0x1021;          // x16+x12+x5+1
+                _width = 16;
+                break;
+
+            case CRC_16BIT_CCITT:
+                _polynomial = 0x1021;          // x16+x12+x5+1
+                _width = 16;
+                _inital_value = ~(0x0);
+                break;
+
+            case CRC_16BIT_IBM:
+                _polynomial = 0x8005;          // x16+x15+x2+1
+                _width = 16;
+                _reflect_data = true;
+                _reflect_remainder = true;
+                break;
+
+            case CRC_32BIT:
+                _polynomial = 0x04C11DB7;     // x32+x26+x23+x22+x16+x12+x11+x10+x8+x7+x5+x4+x2+x+1
+                _width = 32;
+                _inital_value = ~(0x0);
+                _final_xor  = ~(0x0);
+                _reflect_data = true;
+                _reflect_remainder = true;
+                break;
+        }
+    }
+
+    virtual ~FastCRC() { }
+
+    /** Initialize a CRC module, generate CRC static table
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int32_t init() {
+        T remainder;
+        uint32_t polynomial = (_width < 8) ? (_polynomial << (8 - _width)) : _polynomial;
+
+        for (int32_t dividend = 0; dividend < 256; ++dividend) {
+            if ( _width < 8) {
+                remainder = dividend;
+            } else {
+                remainder = dividend << (_width - 8);
+            }
+            for (uint8_t bit = 8; bit > 0; --bit) {
+                if (remainder & TOP_BIT(_width)) {
+                    remainder = (remainder << 1) ^ polynomial;
+                } else {
+                    remainder = (remainder << 1);
+                }
+            }
+
+            /*
+             * Store the result into the table.
+             */
+            _crc_table[dividend] = remainder;
+        }
+        return 0;
+    }
+
+    /** Deinitialize a CRC module
+     *
+     * @return  0 on success, negative error code on failure
+     */
+    virtual int32_t deinit() {
+        return 0;
+    }
+
+    /** Get CRC polynomial
+     *
+     * Get the current CRC polynomial
+     *
+     * @return  Polynomial value
+     */
+    virtual uint32_t get_polynomial(void) const {
+        return _polynomial;
+    }
+
+    /** Compute CRC for the data input as per the polynomial set
+     *
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  crc  CRC
+     *  @return  0  on success or a negative error code on failure
+     *  @note CRC calculations are done byte wise.
+     */
+    virtual int32_t compute(void *buffer, data_size_t size, uint32_t *crc) {
+        MBED_ASSERT(crc != NULL);
```

Comment:
Isn't MBED_ASSERTs removed for non-debug builds? In that case its better to return an error when you get an invalid parameter instead of working with invalid data(works for release builds).

Reply:
> In that case its better to return an error when you get an invalid parameter

== more code size

In this case we'll hit a hard fault either way, which is probably clear enough as to what went wrong IMO.

Reply:
As of now hitting Hard Fault will put the application into an unrecoverable state, where as if we return an error the application can handle that more gracefully, like logging/reporting the error to the user. And hitting hardfault is not a scenario easily debuggable in a customer/deployed scenario and without a debugger. But I do agree that returning error code will slightly increase the code size.

Reply:
But this case is a usage error. The correct way to handle this error is to not pass in NULL in the first place.

Point taken for making it easier to debug in the customer scenario.

But I have a hard time imagining the case where someone propagates error codes correctly, but passes NULL into this function.

---

Just food for thought, I don't think this conversation should block this pr